cmd/mesh/upgrade.go

```diff
@@ -0,0 +1,391 @@
+			return err
+		}
+
+		// Read the upgraded Istio version from the the cluster
+		upgradeVer, err := retrieveControlPlaneVersion(kubeClient, istioNamespace, l)
+		if err != nil {
+			l.logAndPrintf("Failed to read the upgraded Istio version. Error: %v", err)
+			return err
+		}
+
+		l.logAndPrintf("Success. Now the Istio control plane is running at version %v.", upgradeVer)
+		return nil
+	}
+	l.logAndPrintf("Upgrade submitted. Please use `istioctl version` to check the current versions.")
+	return nil
+}
+
+// checkUpgradeValues checks the upgrade eligibility by comparing the current values with the target values
+func checkUpgradeValues(curValues string, tarValues string, l *logger) {
+	diff := compare.YAMLCmp(curValues, tarValues)
+	if diff == "" {
+		l.logAndPrintf("Upgrade check: Values unchanged. The target values are identical to the current values.\n")
+	} else {
+		l.logAndPrintf("Upgrade check: Warning!!! the following values will be changed as part of upgrade. "+
+			"If you have not overridden these values, they will change in your cluster. Please double check they are correct:\n%s", diff)
+	}
+}
+
+// waitForConfirmation waits for user's confirmation if yes is not set
+func waitForConfirmation(yes bool, l *logger) {
+	if yes {
+		return
+	}
+	if !confirm("Confirm to proceed [y/N]?", os.Stdout) {
+		l.logAndFatalf("Abort.")
+	}
+}
+
+// readValuesFromInjectorConfigMap reads the values from the config map of sidecar-injector.
+func readValuesFromInjectorConfigMap(kubeClient manifest.ExecClient, istioNamespace string) (string, error) {
+	configMapList, err := kubeClient.ConfigMapForSelector(istioNamespace, "istio=sidecar-injector")
+	if err != nil || len(configMapList.Items) == 0 {
+		return "", fmt.Errorf("failed to retrieve sidecar-injector config map: %v", err)
+	}
+
+	jsonValues := ""
+	foundValues := false
+	for _, item := range configMapList.Items {
+		if item.Name == "istio-sidecar-injector" && item.Data != nil {
+			jsonValues, foundValues = item.Data["values"]
+			if foundValues {
+				break
+			}
+		}
+	}
+
+	if !foundValues {
+		return "", fmt.Errorf("failed to find values in sidecar-injector config map: %v", configMapList)
+	}
+
+	yamlValues, err := yaml.JSONToYAML([]byte(jsonValues))
+	if err != nil {
+		return "", fmt.Errorf("jsonToYAML failed to parse values:\n%v\nError:\n%v", yamlValues, err)
+	}
+
+	return string(yamlValues), nil
+}
+
+// checkSupportedVersions checks if the upgrade cur -> tar is supported by the tool
+func checkSupportedVersions(cur, tar, versionsURI string, l *logger) error {
+	if cur == tar {
+		return fmt.Errorf("the current version %v equals to the target version %v", cur, tar)
+	}
+
+	curPkgVer, err := pkgversion.NewVersionFromString(cur)
+	if err != nil {
+		return fmt.Errorf("incorrect version: %v", cur)
+	}
+
+	tarPkgVer, err := pkgversion.NewVersionFromString(tar)
+	if err != nil {
+		return fmt.Errorf("incorrect version: %v", tar)
+	}
+
+	if curPkgVer.Major != tarPkgVer.Major {
```

Comment:
sorry, i should've been more explicit before in my last comment - we don't want this logic in here, everything should be driven from the constraints in the version compatibility table. 
we do actually support at least minor version upgrade (it depends on patch version we're coming from). so the table should cover all cases without any built in logic.